### PR TITLE
Reuse an already successful proxy on next requests

### PR DESCRIPTION
**Pull request by benallamar**
RequestHedgingProxyProvider doesn't reuse the successful proxy to send
the next request, but instead it broadcast the latter to all namenode.

### CODE REVIEW

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/ha/RequestHedgingProxyProvider.java [hunk 1]
```diff
@@ -141,10 +137,42 @@ public Object call() throws Exception {
...
+    /**
+     * Set next proxy
+     *
+     * It does return true if the method succeed in setting the next target proxy,
+     * otherwise it returns false.
+     *
+     * @return boolean
+     */
+    private boolean setNextProxy(){
+      targetProxies.remove(toIgnore);
```
Comment (Willymontaz) [inline hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/ha/RequestHedgingProxyProvider.java]: keep this line where it was

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/ha/RequestHedgingProxyProvider.java [hunk 2]
```diff
@@ -85,12 +85,8 @@ public RequestHedgingInvocationHandler(
       try {
         // Optimization : if only 2 proxies are configured and one had failed
```
Comment (Willymontaz) [inline hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/ha/RequestHedgingProxyProvider.java]: Those comments lose their meanings, move them where needed

**Review by Willymontaz — COMMENTED** [review]
As discusses, 
 - mention in your comment the fix with unwrapping the exception
 - I find setNextProxy and nextProxy not relevant names.

**Comment by Willymontaz** [comment]
LGTM

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/ha/RequestHedgingProxyProvider.java [hunk 3]
```diff
@@ -83,14 +83,11 @@ public RequestHedgingInvocationHandler(
...
-        // Optimization : if only 2 proxies are configured and one had failed
-        // over, then we dont need to create a threadpool etc.
         targetProxies.remove(toIgnore);
-        if (targetProxies.size() == 1) {
-          ProxyInfo<T> proxyInfo = targetProxies.values().iterator().next();
-          Object retVal = method.invoke(proxyInfo.proxy, args);
-          successfulProxy = proxyInfo;
-          return retVal;
+
+        ProxyInfo<T> targetProxy;
```
Comment (mlachkar) [inline hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/ha/RequestHedgingProxyProvider.java]: ProxyInfo<T> targetProxy = getSuccessfulOrLeftProxy();

no ?

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/ha/RequestHedgingProxyProvider.java [hunk 4]
```diff
@@ -141,8 +138,45 @@ public Object call() throws Exception {
...
+       * @param method
+       * @param args
+       *
+       * @throws Exception
+       */
+      private Object
+      invokeMethodOnGivenProxy(ProxyInfo<T> proxyInfo, Method method,
+                              final Object[] args) throws Exception{
+           try{
+               return method.invoke(proxyInfo.proxy,args);
```
Comment (mlachkar) [inline hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/ha/RequestHedgingProxyProvider.java]: add a space please before args

**Review by mlachkar — CHANGES_REQUESTED** [review]

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/ha/RequestHedgingProxyProvider.java (hunk 3 above)
Comment (benallamar) [reply hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/ha/RequestHedgingProxyProvider.java]: I do an assignment and check at the same time in the next line

File: hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/ha/RequestHedgingProxyProvider.java (hunk 4 above)
Comment (benallamar) [reply hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/ha/RequestHedgingProxyProvider.java]: DONE

**Review by mlachkar — APPROVED** [review]

**Pull request merged**